cmake/Modules/LibAddHaskellPlugin.cmake

```diff
@@ -40,272 +40,311 @@ macro (add_haskell_plugin target)
-						find_library (GHC_GMP_LIB "HS${GHC_GMP_NAME}${GHC_DYNAMIC_SUFFIX}" PATHS ${GHC_GMP_PATH})
-
-						execute_process (COMMAND ${GHC-PKG_EXECUTABLE} latest ghc-prim
-								 OUTPUT_VARIABLE GHC_PRIM_NAME
-								 OUTPUT_STRIP_TRAILING_WHITESPACE)
-						set (GHC_PRIM_PATH "${GHC_LIB_DIR}/${GHC_PRIM_NAME}")
-						find_library (GHC_PRIM_LIB "HS${GHC_PRIM_NAME}${GHC_DYNAMIC_SUFFIX}" PATHS ${GHC_PRIM_PATH})
-
-						if (GHC_RTS_LIB)
-							if (GHC_BASE_LIB)
-								if (GHC_GMP_LIB)
-									if (GHC_PRIM_LIB)
-
-										set (
-											PLUGIN_HASKELL_NAME
-											"${CMAKE_CURRENT_BINARY_DIR}/libHS${target}${GHC_DYNAMIC_SUFFIX}${GHC_DYNAMIC_ENDING}"
-											)
-
-										set (GHC_LIBS
-										     ${GHC_RTS_LIB}
-										     ${GHC_BASE_LIB}
-										     ${GHC_GMP_LIB}
-										     gmp
-										     ${GHC_PRIM_LIB}
-										     ${PLUGIN_HASKELL_NAME})
-
-										# GHC's structure differs between OSX and Linux
-										# On OSX we need to link iconv and Cffi additionally
-										if (APPLE)
-											find_library (GHC_FFI_LIB
-												      Cffi
-												      PATHS "${GHC_LIB_DIR}/rts")
-											if (GHC_FFI_LIB)
-												set (GHC_LIBS
-												     ${GHC_LIBS}
-												     ${GHC_FFI_LIB}
-												     iconv)
-											else (GHC_FFI_LIB)
-												remove_plugin (${target}
-													       "GHC_FFI_LIB not found")
-											endif (GHC_FFI_LIB)
-										endif (APPLE)
-
-										# configure include paths
-										configure_file (
-											"${CMAKE_CURRENT_SOURCE_DIR}/${target}.cabal.in"
-											"${CMAKE_CURRENT_BINARY_DIR}/${target}.cabal"
-											@ONLY) # configure the haskell plugin base file for
-											       # the current plugin
-										configure_file (
-											"${CMAKE_SOURCE_DIR}/src/plugins/haskell/haskell.c.in"
-											"${CMAKE_CURRENT_BINARY_DIR}/haskell.c"
-											@ONLY) # same for the header
-										configure_file (
-											"${CMAKE_SOURCE_DIR}/src/plugins/haskell/haskell.h.in"
-											"${CMAKE_CURRENT_BINARY_DIR}/haskell.h"
-											@ONLY) # copy the readme so the macro in haskell.c
-											       # finds it
-										file (COPY
-										      "${CMAKE_CURRENT_SOURCE_DIR}/README.md"
-										      DESTINATION
-										      "${CMAKE_CURRENT_BINARY_DIR}") # same for the setup
-														     # logic,  depending on
-														     # wheter a custom  one
-														     # exists  use the
-														     # default  suitable for
-														     # almost  everything
-										set (CABAL_CUSTOM_SETUP_FILE
-										     "${CMAKE_CURRENT_SOURCE_DIR}/Setup.hs.in")
-										if (NOT EXISTS ${CABAL_CUSTOM_SETUP_FILE})
+		if (NOT BUILD_STATIC AND NOT BUILD_FULL)
+			if (NOT ENABLE_ASAN)
+				if (PLUGINPROCESS_FOUND)
+					if (HASKELL_FOUND)
+						check_binding_included ("haskell" HAVE_HASKELL_BINDING)
+						if (HAVE_HASKELL_BINDING)
+
```

Comment:
please put this code in a macro, it is kind of unreadable that way (too deeply nested)